web/packages/teleterm/src/ui/services/clusters/clustersService.ts

```diff
@@ -640,3 +640,24 @@ export function makeApp(source: tsh.App): App {
 
   return { ...source, addrWithProtocol };
 }
+
+/** Returns a URL that can be used to open the app in a browser. */
+export function getWebAppLaunchUrl(
+  app: tsh.App,
+  clustersService: ClustersService
+): string {
+  const { fqdn, publicAddr } = app;
+
+  const rootCluster = clustersService.findCluster(
+    routing.ensureRootClusterUri(app.uri)
+  );
+  const cluster = clustersService.findClusterByResource(app.uri);
+
+  const canCreateUrl =
+    rootCluster.proxyHost && fqdn && cluster?.name && publicAddr;
```

Comment:
I don't fully understand this check in the Web UI. I think it's only there so that it's not possible to generate launch URLs to SAML apps and others, since [the Web UI never checks if `launchUrl` is empty](https://github.com/gravitational/teleport/blob/4a0b7336322e7440f6c5a6fc79c76858efd84e94/web/packages/teleport/src/UnifiedResources/ResourceActionButton.tsx#L195).

With that said, I think we should not check `rootCluster.proxyHost`, as we should be able to assume that it's always present. Same with `cluster.name` at this point in the program I think.

The check should emphasize the fact that the launch URL can be generated only for regular apps, not SAML etc. The Web UI doesn't express this clearly, but if I understand the purpose of this check correctly, we have a chance to stop copying this defensive approach in teleterm.